pkg/mappings.bzl

```diff
@@ -0,0 +1,403 @@
+            (`strip_prefix.files_only()`).
+            
+            Note that this only functions on paths that are known at analysis
+            time.  Specifically, this will not consider directories within
+            TreeArtifacts (directory outputs), or the directories themselves.
+            See also #269.
+            """,
+            default = strip_prefix.files_only(),
+        ),
+        "excludes": attr.label_list(
+            doc = """List of files or labels to exclude from the inputs to this rule.
+
+            Mostly useful for removing files from generated outputs or
+            preexisting `filegroup`s.
+            """,
+            default = [],
+            allow_files = True,
+        ),
+        "renames": attr.label_keyed_string_dict(
+            doc = """Destination override map.
+
+            This attribute allows the user to override destinations of files in
+            `pkg_file`s relative to the `prefix` attribute.  Keys to the
+            dict are source files/labels, values are destinations relative to
+            the `prefix`, ignoring whatever value was provided for
+            `strip_prefix`.
+
+            This is the most effective way to rename files using `pkg_file`s.
+
+            The following keys are rejected:
+
+            - Any label that expands to more than one file (mappings must be
+              one-to-one).
+
+            - Any label or file that was either not provided or explicitly
+              `exclude`d.
+            """,
+            default = {},
+            allow_files = True,
+        ),
+    },
+)
+
+def _pkg_mkdirs_impl(ctx):
+    _validate_attr(ctx.attr.attrs)
+
+    return [
+        PackageDirsInfo(
+            dirs = ctx.attr.dirs,
+            attributes = ctx.attr.attrs,
+        ),
+    ]
+
+pkg_mkdirs = rule(
+    doc = """Defines creation and ownership of directories in packages
+
+    Use this if:
+
+    1) You need to create an empty directory in your package.
+
+    2) Your package needs to explicitly own a directory, even if it already owns
+       files in those directories.
+
+    3) You need nonstandard permissions (typically, not "0755") on a directory
+       in your package.
+
+    For some package management systems (e.g. RPM), directory ownership (2) may
+    imply additional semantics.  Consult your package manager's and target
+    distribution's documentation for more details.
+    """,
+    implementation = _pkg_mkdirs_impl,
+    # @unsorted-dict-items
+    attrs = {
+        "dirs": attr.string_list(
+            doc = """Directory names to make within the package
+
+            If any part of the requested directory structure does not already
+            exist within a package, the package builder will create it for you
+            with a reasonable set of default permissions (typically `0755
+            root.root`).
+
+            """,
+            mandatory = True,
+        ),
+        "attrs": attr.string_list_dict(
```

Comment:
I wonder if this should be called "properties", so that people don't have to think about the "attributes attribute".

This can be done later, perhaps.